### PR TITLE
Bump Fourmolu to 0.4

### PR DESCRIPTION
**Pull request by georgefst**
Addresses the relevant checklist item in #297.

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2254"></a>

### CODE REVIEW

File: plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs [hunk 1]
```diff
@@ -8,19 +8,19 @@ module Ide.Plugin.Fourmolu (
...
 import           Control.Monad.IO.Class
-import           Data.Bifunctor              (first)
-import qualified Data.Text                   as T
-import           Development.IDE             hiding (pluginHandlers)
-import           Development.IDE.GHC.Compat  as Compat hiding (Cpp)
+import           Data.Bifunctor                  (first)
+import qualified Data.Text                       as T
+import           Development.IDE                 hiding (pluginHandlers)
+import           Development.IDE.GHC.Compat      as Compat hiding (Cpp)
 import qualified Development.IDE.GHC.Compat.Util as S
```
Comment (georgefst) [inline plugins/hls-fourmolu-plugin/src/Ide/Plugin/Fourmolu.hs:18]: I don't know how this file was passing CI before? Seeing as this line causes `stylish-haskell` to reformat all the other imports (I personally detest this behaviour, but that's an argument for another day).

File: plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal [hunk 2]
```diff
@@ -22,7 +22,7 @@ library
   build-depends:
     , base            >=4.12 && <5
     , filepath
-    , fourmolu        ^>=0.3
+    , fourmolu        ^>=0.3 || ^>=0.4
```
Comment (georgefst) [inline plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal:25]: Ideally, I'd remove 0.3 entirely, as 0.4 includes some important bug fixes. But this isn't possible while `stylish-haskell` is tied to older versions of GHC and Cabal.

(Same applies to Ormolu here - right now, the latest version will never be chosen by the dependency solver)

**Comment by jneira** [comment]
@georgefst many thanks to spend time on https://github.com/fourmolu/fourmolu/issues/104 and this, getting support for ghc-9

File: plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal (hunk 2 above)
Comment (georgefst) [reply plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal:25]: Actually, looks like this only applies to GHC < 9. The `stylish-haskell` plugin is disabled for 9.0.

**Comment by jneira** [comment]
I hope you dont mind i enabled fourmolu for ghc-9 to give it a shot inci

File: plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal (hunk 2 above)
Comment (jneira) [reply plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal:25]: Maybe adding a comment about, to remove 0.3 some day?

**Comment by georgefst** [comment]
> @georgefst many thanks to spend time on [fourmolu/fourmolu#104](https://github.com/fourmolu/fourmolu/issues/104) and this, getting support for ghc-9

Thanks, but @brandonchinn178 has actually done most of the hard work recently.

**Review by jneira — APPROVED** [review]
lgtm i hope fourmolu will continue rocking in the formatter side

**Comment by jneira** [comment]
There is a build error for ghc-9:

```
 src/Ide/Plugin/Fourmolu.hs:20:1: error:
Error:     Could not find module ‘GhcPlugins’
    Perhaps you meant GHC.Plugins (from ghc-9.0.1)
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
   |
20 | import           GhcPlugins                      (HscEnv (hsc_dflags))
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by georgefst** [comment]
> There is a build error for ghc-9:
> 
> ```
>  src/Ide/Plugin/Fourmolu.hs:20:1: error:
> Error:     Could not find module ‘GhcPlugins’
>     Perhaps you meant GHC.Plugins (from ghc-9.0.1)
>     Use -v (or `:set -v` in ghci) to see a list of the files searched for.
>    |
> 20 | import           GhcPlugins                      (HscEnv (hsc_dflags))
>    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ```

That import should have been removed in ec53fcb7b54434d62901a73bbf3e3085a579c487. Fixed in 46d3264.

This would have been caught by `-Wall`, but for some reason that isn't enabled for many of the plugins?

File: plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal (hunk 2 above)
Comment (georgefst) [reply plugins/hls-fourmolu-plugin/hls-fourmolu-plugin.cabal:25]: > Maybe adding a comment about, to remove 0.3 some day?

Actually, I think there's no need. I'll probably end up at least trying to remove it every time I bump the upper bound, and eventually it'll be possible.

**Comment by jneira** [comment]
> This would have been caught by -Wall, but for some reason that isn't enabled for many of the plugins?

Jumm they should afaiu, maybe is cargo culting and it is enabled for hls itself, maybe it is time to give a try at least here

**Review by jneira — COMMENTED** [review]
what about `ghc-options:      -Wall -Wredundant-constraints`, like hls itself?

(without `-Wno-name-shadowing -Wno-unticked-promoted-constructors` if this does not need it)

**Comment by georgefst** [comment]
> `-Wredundant-constraints`

Meh, I think there are legitimate use cases for redundant constraints, though it's very unlikely to come up in this plugin anyway.
 
> `-Wno-name-shadowing`

I prefer to be warned

> `-Wno-unticked-promoted-constructors`

No strong opinion, but again unlikely to come up.

On the whole, I'd go for keeping things simple with just `-Wall`. But I don't feel strongly about any of it.

**Comment by jneira** [comment]
> 
> 
> > `-Wredundant-constraints`
> 
> Meh, I think there are legitimate use cases for redundant constraints, though it's very unlikely to come up in this plugin anyway.
> 
> > `-Wno-name-shadowing`
> 
> I prefer to be warned.
> 
> > `-Wno-unticked-promoted-constructors`
> 
> No strong opinion, but again unlikely to come up.
> 
> On the whole, I'd go for keeping things simple with just `-Wall`. But I don't feel that strongly about any of it.

Agree with you, i forgot to mention `-Wincomplete-uni-patterns` (it is in ghcide but no hls, ew *have* to unify all falgs :sweat:)

**Comment by georgefst** [comment]
> it is in ghcide but no hls, we _have_ to unify all flags for sure 😓

Yeah, the fact that I know of no good way to do this in a multi-cabal-package project (`dhall-to-cabal` with flags etc. defined in a single file? probably a totally uncharted path and almost certainly more effort than it's worth right now) is why I suggest keeping it simple.

**Comment by georgefst** [comment]
CI test failure looks unrelated to Fourmolu...

**Comment by jneira** [comment]
> 
> 
> CI test failure looks unrelated to Fourmolu...

Yeah, all pr's are affected by #2253 :cry:

**Comment by georgefst** [comment]
> > CI test failure looks unrelated to Fourmolu...
> 
> Yeah, all pr's are affected by #2253 😢

Ah, ok.

**Comment by jneira** [comment]
> 
> 
> > > CI test failure looks unrelated to Fourmolu...
> > 
> > 
> > Yeah, all pr's are affected by #2253 cry
> 
> Ah, ok.

I was wrong, not all pr's are affected only those which are updating the cabal.project index state so maybe a new dep version or the cache invalidation is triggering it

**Pull request merged**